nni/contrib/compression/pruning/movement_pruner.py

```diff
@@ -0,0 +1,226 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+from __future__ import annotations
+
+from collections import defaultdict
+import logging
+from typing import Dict, List, overload
+
+import torch
+from torch.optim import Adam
+
+from .scheduled_pruner import ScheduledPruner
+from .tools import is_active_target, generate_sparsity
+from ..base.compressor import Compressor
+from ..base.target_space import TargetType
+from ..base.wrapper import ModuleWrapper
+from ..utils import Evaluator
+
+MOVEMENT_SCORE_PNAME = '{}_mvp_score'
+_logger = logging.getLogger(__name__)
+
+
+class MovementPruner(ScheduledPruner):
+    """
+    Movement pruner is an implementation of movement pruning.
+    This is a "fine-pruning" algorithm, which means the masks may change during each fine-tuning step.
```

Comment:
"fine-pruning" is the word that you coined?

Reply:
no, https://proceedings.neurips.cc/paper/2020/file/eae15aabaa768ae4a5993a8a4f4fa6e4-Paper.pdf

In introduction:
```
While magnitude pruning is highly effective for standard supervised learning, it is inherently less
useful in the transfer learning regime. In supervised learning, weight values are primarily determined
by the end-task training data. In transfer learning, weight values are mostly predetermined by the
original model and are only fine-tuned on the end task. This prevents these methods from learning to
prune based on the fine-tuning step, or “fine-pruning.”
```